### PR TITLE
fix: [CDS-78563] Replace release identifier for instance sync if already exists

### PR DESCRIPTION
**Pull request by alebuh**
## Describe your changes
When HELM_STEADY_STATE_CHECK_1_16 is disabled containerServiceName will have actual value and the release identifier will look something like this:
ReleaseIdentifert1( releaseName: release1, namespace: namespace1, containerServiceName: workloadName )
When HELM_STEADY_STATE_CHECK_1_16  is enabled containerServiceName will be null and look like this:
ReleaseIdentifert2( releaseName: release1, namespace: namespace1, containerServiceName: null )
After merging release identifiers only one should be present as they are part of the same release. We always should use the latest one
## Test Case
Only for V2
1. The first deployment HELM_STEADY_STATE_CHECK_1_16 is disabled, and the second is enabled. InstanceSync doesn't make duplicates. (Each instance counts once)
2. The first deployment HELM_STEADY_STATE_CHECK_1_16 is enabled, and the second is disabled. InstanceSync doesn't make duplicates. (Each instance counts once)

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54671)

### CODE REVIEW

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger messagecheck

File: 400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2DeploymentHelperCg.java [hunk 1]
```diff
@@ -313,4 +312,21 @@ public long getDeleteReleaseAfter(CgReleaseIdentifiers releaseIdentifier, Instan
...
+      Set<CgReleaseIdentifiers> identifiers, CgK8sReleaseIdentifier k8sNewIdentifier) {
+    // When HELM_STEADY_STATE_CHECK_1_16 is enabled containerServiceName will be null, but when it's disabled it will
+    // have actual value. We should ignore this filed as they are part of the same release, and should always use new
+    // identifier
+    for (CgReleaseIdentifiers identifier : identifiers) {
+      CgK8sReleaseIdentifier cgK8sReleaseIdentifier = (CgK8sReleaseIdentifier) identifier;
+      if (k8sNewIdentifier.getReleaseName().equals(cgK8sReleaseIdentifier.getReleaseName())
+          && k8sNewIdentifier.getNamespace().equals(cgK8sReleaseIdentifier.getNamespace())
+          && k8sNewIdentifier.getClusterName().equals(cgK8sReleaseIdentifier.getClusterName())
+          && k8sNewIdentifier.isHelmDeployment() == (cgK8sReleaseIdentifier.isHelmDeployment())) {
```
Comment (abhinavsingh1196) [inline 400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2DeploymentHelperCg.java:326]: will checkstyle or PMD check complain about the redundant braces around `cgK8sReleaseIdentifier.isHelmDeployment()`?

Comment (abhinavsingh1196) [reply 400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2DeploymentHelperCg.java:326]: I see its passed :)

**Review by abhinavsingh1196 — APPROVED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54671)* status: 2 file(s) to review. 1 of 2 approvals obtained.

**Review by TarunUba1 — APPROVED** [review]

File: 400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2DeploymentHelperCg.java [hunk 2]
```diff
@@ -313,4 +312,21 @@ public long getDeleteReleaseAfter(CgReleaseIdentifiers releaseIdentifier, Instan
...
+  private static void addIdentifierAndReplaceExistingIfAlreadyExist(
+      Set<CgReleaseIdentifiers> identifiers, CgK8sReleaseIdentifier k8sNewIdentifier) {
+    // When HELM_STEADY_STATE_CHECK_1_16 is enabled containerServiceName will be null, but when it's disabled it will
+    // have actual value. We should ignore this filed as they are part of the same release, and should always use new
+    // identifier
+    for (CgReleaseIdentifiers identifier : identifiers) {
+      CgK8sReleaseIdentifier cgK8sReleaseIdentifier = (CgK8sReleaseIdentifier) identifier;
+      if (k8sNewIdentifier.getReleaseName().equals(cgK8sReleaseIdentifier.getReleaseName())
+          && k8sNewIdentifier.getNamespace().equals(cgK8sReleaseIdentifier.getNamespace())
+          && k8sNewIdentifier.getClusterName().equals(cgK8sReleaseIdentifier.getClusterName())
```
Comment (namantalaycha10) [inline 400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2DeploymentHelperCg.java:325]: please do the null check before doing equal operation

**Review by namantalaycha10 — CHANGES_REQUESTED** [review]

File: 400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2DeploymentHelperCg.java (hunk 2 above)
Comment (alebuh) [reply 400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2DeploymentHelperCg.java]: done

**Comment by alebuh** [comment]
trigger uts

**Review by abhinavsingh1196 — DISMISSED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54671)* status: 2 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. Waiting on team(s): k8s-team.

**Comment by alebuh** [comment]
trigger uts

**Review by TarunUba1 — APPROVED** [review]

**Review by GoelPratyush99 — APPROVED** [review]

**Review by acasian-harness — APPROVED** [review]

**Review by abosii-harness — APPROVED** [review]

**Review by namantalaycha10 — APPROVED** [review]
lgtm

**Pull request merged**